src/csharp/Grpc.Core.Tests/ContextualMarshallerTest.cs

```diff
@@ -0,0 +1,119 @@
+#region Copyright notice and license
+
+// Copyright 2018 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Grpc.Core;
+using Grpc.Core.Internal;
+using Grpc.Core.Utils;
+using NUnit.Framework;
+
+namespace Grpc.Core.Tests
+{
+    public class ContextualMarshallerTest
+    {
+        const string Host = "127.0.0.1";
+
+        MockServiceHelper helper;
+        Server server;
+        Channel channel;
+
+        [SetUp]
+        public void Init()
+        {
+            var contextualMarshaller = new Marshaller<string>(
+                (str, serializationContext) =>
+                {
+                    if (str == "UNSERIALIZABLE_VALUE")
+                    {
+                        // Google.Protobuf throws exception inherited from IOException
+                        throw new IOException("Error serializing the message.");
+                    }
+                    if (str == "SERIALIZE_TO_NULL")
+                    {
+                        return;
+                    }
+                    var bytes = System.Text.Encoding.UTF8.GetBytes(str);
+                    serializationContext.Complete(bytes);
```

Comment:
Will the `SerializationContext` interface have other methods besides `Complete`?

If not, I'm wondering why not keep the current `return`-based API?

Reply:
with the return-based API, one basically has to allocate a new object (buffer) each time a message is serialized, but for performance reasons we don't want that. The plan is to add more methods that will allow to serialize messages directly to native memory (a grpc slice) so that we don't have to copy any data unnecessarily and to allow gRPC c# library to use different memory management strategies than allocating one-time buffers.
The new apis will probably have more methods e.g. `GetMemory(int sizeHint)` to request a memory to serialize into  and then a parameterless `Complete()` method to announce that we are done with serialization. (this is similar idea as PipeWriter with GetMemory(), Advance() and Flush() methods https://github.com/dotnet/corefx/blob/master/src/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriter.cs)